kytos/core/link.py

```diff
@@ -130,56 +131,50 @@ def available_tags(self):
         return [tag for tag in self.endpoint_a.available_tags if tag in
                 self.endpoint_b.available_tags]
 
-    def use_tag(self, tag):
-        """Remove a specific tag from available_tags if it is there.
-
-        Deprecated: use only the get_next_available_tag method.
-        """
-        if self.is_tag_available(tag):
-            self.endpoint_a.use_tag(tag)
-            self.endpoint_b.use_tag(tag)
-            return True
-        return False
-
     def is_tag_available(self, tag):
         """Check if a tag is available."""
         return (self.endpoint_a.is_tag_available(tag) and
                 self.endpoint_b.is_tag_available(tag))
 
-    def get_next_available_tag(self):
+    def get_next_available_tag(self, controller, tag_type: str = '1') -> TAG:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
             # Copy the available tags because in case of error
             # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
+            available_tags_a = deepcopy(self.endpoint_a.get_available_tags())
```

Comment:
I believe this deepcopy can get avoided, you've also contributed to link having deterministic endpoints before, so to also protect from race conditions, a deepcopy should be needed if can acquire the following locks:

```
with self._get_available_vlans_lock[link.id]: 
  with self.endpoint_a._tag_lock:
    with self.endpoint_b._tag_lock:
       ...
```       
       
I'd expect this to be deadlock free too, if you can try it out. Notice that you'll also need to refactor `self._get_available_vlans_lock` be a `dict: [str, Lock]`, this is an instance attr, since mef_eline is instantiating links, then it doesn't share the same instance lock, so we'll need to keep it at the instance level for now, until links are moved to the core, but it'll also need to take a link.id, otherwise it'll hold a lock for too long.

Reply:
Looking into this too.

Reply:
Added locks for each endpoint. `_get_available_vlans_lock` is maintained as `Lock` since it is called inside a link.

Reply:
Check out this comment [here](https://github.com/kytos-ng/kytos/pull/409#discussion_r1323596362) I further expanded on the locks comment . The current "triple" one isn't acquiring the same ones that suggested, I elaborated on the comment explaining what I meant, let me know if it makes sense or if you have questions. We could also managed to pull it off without a triple lock but keeping a shallow copy, but the triple one will make things safer and less surprising to deal with I'd say.

Reply:
Triple lock updated

Reply:
I followed up here https://github.com/kytos-ng/kytos/pull/409#discussion_r1329328675, check it out